static-site/components/list-resources/listResourcesApi.ts

```diff
@@ -0,0 +1,335 @@
+import { ResourceType, Resource, Group, PathVersionsForGroup, FetchGroupHistory } from "./types";
+import { InternalError } from "../error-boundary";
+
+interface ResourceListingDatasets {
+  pathPrefix: string;
+  pathVersions: Record<string, string[]>;
+}
+
+interface ResourceListingIntermediates {
+  pathPrefix: string;
+  latestVersions: {
+    [id: string]: {
+      [filename: string]: [url: string, mostRecentlyIndexed: string]
+    }
+  }
+}
+
+/**
+ * A callback function used by the <ListResources> component to fetch information
+ * about the resources to list from the host server's /list-resources API endpoint.
+ */
+export async function listResourcesAPI(
+  sourceId: string,
+  resourceType: ResourceType,
+  {versioned, groupDisplayNames, groupUrl}: {
+    /** Report prior versions of each resource.
+     * TODO: infer this from the API data itself
+     */
+    versioned: boolean,
+    groupDisplayNames?: Record<string, string>,
+    groupUrl?: (groupName: string) => string
+  }
+): Promise<Group[]> {
+  const requestPath = `/list-resources/${sourceId}/${resourceType}`;
+  const response = await fetch(requestPath, {
+    headers: { accept: "application/json" },
+  });
+  if (response.status !== 200) {
+    throw new Error(
+      `fetching application/json data from "${requestPath}" returned status code ${response.status}`,
+    );
+  }
+  // eslint-disable-next-line @typescript-eslint/consistent-type-assertions
+  const data = ((await response.json())[resourceType][sourceId]) as ResourceListingDatasets |  ResourceListingIntermediates;
+  const urlBuilder = (name) => `/${data.pathPrefix}/${name}`;
+  const areDatasets = (x: ResourceListingDatasets |  ResourceListingIntermediates): x is ResourceListingDatasets => {
+    return Object.hasOwn(x, 'pathVersions');
+  }
```

Comment:
Strong, non-blocking preference: use a function unless there's an actual reason for storing an anonymous function in a variable. (AFAICT, here, there isn't.)